core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java

```diff
@@ -0,0 +1,255 @@
+            throw new UnsupportedOperationException("Policy not supported: " + thresholdPolicy);
+      }
+      DefaultEvictionListener evictionListener = new DefaultEvictionListener();
+      evictionCache = applyListener(caffeine, evictionListener, new SegmentMapUpdater()).build();
+      entries = evictionCache.asMap();
+   }
+
+   public BoundedSegmentedDataContainer(int numSegments, long thresholdSize,
+         EntrySizeCalculator<? super K, ? super InternalCacheEntry<K, V>> sizeCalculator) {
+      super(numSegments);
+      DefaultEvictionListener evictionListener = new DefaultEvictionListener();
+
+      evictionCache = applyListener(Caffeine.newBuilder()
+            .weigher((K k, InternalCacheEntry<K, V> v) -> (int) sizeCalculator.calculateSize(k, v))
+            .maximumWeight(thresholdSize), evictionListener, new SegmentMapUpdater())
+            .build();
+
+      entries = evictionCache.asMap();
+   }
+
+   /**
+    * CacheWriter that is invoked only when {@link java.util.Map#remove(Object)} or
+    * {@link java.util.Map#put(Object, Object)} are invoked. We have to handle compute* methods manually.
+    */
+   private class SegmentMapUpdater implements CacheWriter<K, InternalCacheEntry<K, V>> {
+      @Override
+      public void write(K key, InternalCacheEntry<K, V> value) {
+         computeEntryWritten(key, value);
+      }
+
+      @Override
+      public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
+         computeEntryRemoved(key, value);
+      }
+   }
+
+   @Override
+   protected void computeEntryWritten(K key, InternalCacheEntry<K, V> value) {
+      int segment = getSegmentForKey(key);
+      ConcurrentMap<K, InternalCacheEntry<K, V>> map = BoundedSegmentedDataContainer.super.getMapForSegment(segment);
+      if (map != null) {
+         map.put(key, value);
+      }
+   }
+
+   @Override
+   protected void computeEntryRemoved(K key, InternalCacheEntry<K, V> value) {
+      int segment = getSegmentForKey(key);
+      ConcurrentMap<K, InternalCacheEntry<K, V>> map = BoundedSegmentedDataContainer.super.getMapForSegment(segment);
+      if (map != null) {
+         map.remove(key, value);
+      }
+   }
+
+   @Override
+   protected ConcurrentMap<K, InternalCacheEntry<K, V>> getMapForSegment(int segment) {
+      // All writes and other ops go directly to the caffeine cache
+      return entries;
+   }
+
+   // Peek is special so we don't hit the caffeine map for read operations
+   @Override
+   public InternalCacheEntry<K, V> peek(Object k) {
+      return peek(getSegmentForKey(k), k);
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> peek(int segment, Object k) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> map = super.getMapForSegment(segment);
+      if (map != null) {
+         return map.get(k);
+      }
+      return null;
+   }
+
+   @Override
+   public void clear() {
+      entries.clear();
+      // TODO: find out if above clear removes these as well
```

Comment:
Is this to be resolved in this PR? Or future work?

Reply:
I can try real quick.

Reply:
Removed it. Also looked at caffeine code and they do invoke the `CacheWriter`.